File: api/fastfile.php
<?php

/*
// CLIENT EXAMPLE CODE:
// use a remote fastfile key to transfer a local file to a remote destination.
// -------------------------------------------------------------------------------------------------------------------
require_once('./fastfile.php');
try {
	// setup our fastfile client object to send to <destionation> using <key>.
	$client = new Fastfile('HOSTNAME_HERE', 'KEY_HERE');

	// apply any custom/special options to the request. (complete list below)
//	$client	->port(8080)				// alternate HTTP port.
//		->ssl(true)				// use HTTPS. (changes default port to 443, unless specified)
//		->timeout(20)				// give up if idle activity timeout is exceeded. (seconds)
//		->bufSize(16384)			// (write) buffer size.
//		->connections(15)			// number of async connections to send data on.
//		->failedConnections(50)			// give up after X number of failed connections.
//		->ipv(6)				// explicit ipv4 or ipv6 protocol.
//		->location('fastfile')			// alternate mod_fastfile location.
//		->chunkSize(2097152)			// break file into X byte size chunks.
//		->host('alt.domain.com')		// specify explicit "Host" header. (does not effect what is connected to)
//		->header('HTTP-Header', 'data');	// specify arbitrary HTTP headers. (multiple headers allowed)

	// transfer the file to the server.
	$client->send('PATH_TO_LOCAL_FILE_HERE');

	print "Fastfile client successfully transferred file.\n";
}
catch (Exception $e) {
	print "Fastfile client error: ".$e->getMessage()."\n";
}
// -------------------------------------------------------------------------------------------------------------------



// SERVER EXAMPLE CODE:
// make a fastfile key (and associated files) to be handed down to a client via whatever means the programmer desires.
// -------------------------------------------------------------------------------------------------------------------
require_once('./fastfile.php');
try {
	$destFile = 'PATH_TO_DESTINATION_FILE_HERE'; // the file the key is associated with. (where client file will be written)
	$key = Fastfile::server($destFile, 0); // optional 2nd argument to specify the maximum filesize. (0=unlimited)
	print "Key ready to be handed down to the client: $key\n";
}
catch (Exception $e) {
	print "Fastfile::server error: ".$e->getMessage()."\n";
}
// -------------------------------------------------------------------------------------------------------------------
*/

class Fastfile {

	/* imported constants/defines from fastfile.h */
	const FF_CONF_PATH = '/etc/fastfile.conf';
	const FF_POOL_DIR_MODE = 2733;
	const FF_POOL_FILE_MODE = 0660;
	const FF_KEY_CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
	const FF_KEY_LEN = 32;
	const FF_CLIENT = 'ffcp';

	/* ------------------------------------------------------------------------------------------------------------------------------------------------------------------- */
	/* client configuration array that holds all options. */
	private $client = array();

	/* CLIENT METHOD: create a new client instance. */
	public function __construct($hostname, $key) {
		$this->client['hostname'] = $hostname;
		$this->client['key'] = $key;
	}

	/* CLIENT METHOD: a catch-all method to handle all client options. */
	public function __call($method, $args) {
		switch($method) {
			case 'ssl':
			case 'location':
			case 'bufSize':
			case 'timeout':
			case 'chunkSize':
			case 'connections':
			case 'failedConnections':
			case 'host':
			case 'port':
			case 'ipv':
				if(count($args) > 0)
					$this->client[$method] = $args[0];
				return($this);

			/* header option supports multiples so it gets special treatment. */
			case 'header':
				if(empty($this->client['header']) || !is_array($this->client['header']))
					$this->client['header'] = array();
				$this->client['header'][$args[0]] = $args[1];
				return($this);
		}
		throw new Exception("unknown fastfile client method: $method");
	}

	/* CLIENT METHOD: send the file to the remote apache/mod_fastfile server. */
	public function send($file) {

		/* make sure the file exists that we want to send. */
		if(!file_exists($file))
			throw new Exception("could not find file to send: $file");

		/* map class options to ffcp command-line options. */
		$map = array(
			'ssl'			=> ' -s',
			'location'		=> ' -l ',
			'bufSize'		=> ' -b ',
			'timeout'		=> ' -t ',
			'chunkSize'		=> ' -c ',
			'connections'		=> ' -n ',
			'failedConnections'	=> ' -f ',
			'host'			=> ' -h ',
			'port'			=> ' -p ',
			'ipv'			=> ' -',
			'header'		=> ' -H '
		);

		/* setup initial command and add any special options using the map above. */
		$cmd = $this->clientFindFfcp();
		$cmd .= ' -q';
		foreach($this->client as $option => $value) {
			if(empty($map[$option]))
				continue;
			switch($option) {
				case 'header':
					foreach($value as $header => $headerData)
						$cmd .= $map[$option].escapeshellarg("$header=$headerData");
					break;
				case 'ssl':
					if(!empty($value))
						 $cmd .= $map[$option];
					break;
				case 'ipv':
					if(!in_array($value, array('4', '6')))
						break; /* continue into default. */
				default:
					$cmd .= $map[$option].escapeshellarg($value);
					break;
			}
		}

		/* append host, key and the file being sent and the command is ready to go. */
		$cmd .= ' '.escapeshellarg($this->client['hostname']).' '.escapeshellarg($this->client['key']).' '.escapeshellarg($file);

		/* prepare and execute the ffcp process. */
		$pipeInfo = array(
			0 => array('pipe', 'r'),
			1 => array('pipe', 'w'),
			2 => array('pipe', 'w')
		);
		$proc = proc_open($cmd, $pipeInfo, $pipes);
		if(!is_resource($proc))
			throw new Exception('lost '.self::FF_CLIENT.' execution process.');

		/* see if we have an error string to store. */
		$ret = stream_get_contents($pipes[2]);

		/* ffcp reported an error occured. */
		if(proc_close($proc)) {
			if($ret)
				throw new Exception(trim($ret));
			else
				throw new Exception(self::FF_CLIENT.' execution process failed');
		}

		/* if we made it to this point without an exception it's a success. */
		return;
	}

	/* CLIENT METHOD: find the ffcp binary path, abort if not found. */
	private function clientFindFfcp() {
		if(($paths = getenv('PATH'))) {
			foreach(explode(':', $paths) as $path) {
				$testPath = $path . '/' . self::FF_CLIENT;
				if(file_exists($testPath))
					return($testPath);
			}
		}
		throw new Exception("could not find '".self::FF_CLIENT."' binary (included with fastfile) required for transmission, check your \$PATH and make sure ".self::FF_CLIENT." is installed.");
	}


	/* ------------------------------------------------------------------------------------------------------------------------------------------------------------------- */
	/* SERVER METHOD: populate the pool directory with a new key, return the newly created key to be handed off the client by whatever means determined by the programmer. */
	public static function server($path, $maxSize=0) {

		/* arrayize fastfile.conf and make sure it's up to snuff. */
		$conf = self::serverParseConf(self::FF_CONF_PATH);

		/* see if "pool" definition exists in the config file. */
		if(empty($conf['pool']))
			throw new Exception("pool directory not set in configuration file: ".self::FF_CONF_PATH);

		/* make sure the pool dir is set up properly. (throws an exception on failure) */
		self::serverValidatePoolDir($conf['pool']);

		/* see if the destination file exists already and overwrite isn't enabled. */
		if(file_exists($path) && (empty($conf['overwrite']) || strtolower($conf['overwrite']) != 'yes'))
			throw new Exception("file already exists: $path");

		return(self::serverGenerateKey($conf['pool'], $path, $maxSize));
	}

	/* SERVER METHOD: create an automatically generated key (and associated files) on the file system for use by the client. */
	private function serverGenerateKey($pool, $path, $maxSize=0) {

		/* loop until we find a key that isn't in use. */
		do {
			/* prepare our files for this key. */
			$key = self::serverMakeKey();
			$poolKey = "$pool/$key";
			$poolDat = "$poolKey.dat";
			$poolLnk = "$poolKey.lnk";


		} while (file_exists($poolKey) || file_exists($poolDat) || file_exists($poolLnk) || is_link($poolLnk));

		/* attempt to create the data file associated with the new key. */
		if(($fh = fopen($poolDat, 'w')) === false)
			throw new Exception("could not create data file in pool directory: $poolDat");

		/* if specified, make the file the specified size for the client to fill in. */
		if($maxSize > 0 && !ftruncate($fh, $maxSize)) {
			fclose($fh);
			unlink($poolDat);
			throw new Exception("could not create data file with the size desired: $poolDat ($maxSize bytes)");
		}
		fclose($fh);

		/* make the link destination, where the file will go when complete. */
		if(!symlink($path, $poolLnk)) {
			unlink($poolDat);
			throw new Exception("could not create symbolic link to destination: $poolLnk -> $path");
		}

		/* create the pool key itself, to be filled in by client data. */
		if(!touch($poolKey)) {
			unlink($poolDat);
			unlink($poolLnk);
			throw new Exception("could not create key file to destination: $poolKey");
		}

		/* set our permissions accordingly for the key and data file. */
		chmod($poolDat, self::FF_POOL_FILE_MODE);
		chmod($poolKey, self::FF_POOL_FILE_MODE);

		return($key);
	}

	/* SERVER METHOD: generate a new (un-validated) fastfile key. */
	private function serverMakeKey() {
		$key = '';
		for($i=0, $len=strlen(self::FF_KEY_CHARS)-1; $i < self::FF_KEY_LEN; $i++)
			$key .= substr(self::FF_KEY_CHARS, mt_rand(0, $len), 1);
		return($key);
	}

	/* SERVER METHOD: validate that the pool directory is setup correctly. */
	private function serverValidatePoolDir($dir) {

		/* see if the pool directory exists. */
		if(!file_exists($dir))
			throw new Exception("pool directory is inaccessible or does not exist: $dir");

		/* make sure the pooldir has the correct permissions.  (pretty funky way of getting octal mode values) */
		if(substr(sprintf("%o", fileperms($dir)), -4) != self::FF_POOL_DIR_MODE)
			throw new Exception("pool directory has incorrect permissions: $dir (should be mode ".self::FF_POOL_DIR_MODE.")");

		return;
	}

	/* SERVER METHOD: all-puroise fastfile.conf config parser, returns all options or specified option. */
	private function serverParseConf($conf, $option=null) {

		/* see if the conf file exists. */
		if(!file_exists($conf) || ($lines = file($conf, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES)) === false)
			throw new Exception("could not access config file: $conf");

		$ret = array();
		foreach($lines as $line) {

			/* skip empty lines and comments. */
			if(strlen($line) < 1 || strstr('#;', $line[0]))
				continue;

			/* split array up and trim the name and value. */
			$val = array_map('trim', explode('=', $line, 2));

			/* skip empty name=value pairs. */
			if(empty($val[0]) || empty($val[1]))
				continue;

			/* add value to our array of all options. (will overwrite) */
			$ret[$val[0]] = $val[1];
		}

		/* if we're only after a particular option just return that. */
		if(!empty($option)) {
			if(empty($ret[$option]))
				throw new Exception("config file did not contain option: $option ($conf)");

			return($ret[$option]);
		}

		/* ... or just return the whole thing. */
		return($ret);
	}
}

?>

File: api/fastfile.cs
/*
// CLIENT EXAMPLE CODE(using cygwin version ffcp.exe):
// use a remote fastfile key to transfer a local file to a remote destination.
// -------------------------------------------------------------------------------------------------------------------
try {
    // setup our fastfile client object to send to <destionation> using <key>.
    fastfile = new Fastfile(@"\PATH\TO\FFCP\DIR", "HOSTNAME_HERE", "KEY_HERE");

    // apply any custom/special options to the request. (complete list below)
//    fastfile.set("port", 8080);                     // alternate HTTP port.
//    fastfile.set("ssl", true);                      // use HTTPS. (changes default port to 443, unless specified)
//    fastfile.set("timeout", 20);                    // give up if idle activity timeout is exceeded. (seconds)
//    fastfile.set("bufSize", 16384);                 // (write) buffer size.
//    fastfile.set("connections", 15);                // number of async connections to send data on.
//    fastfile.set("failedConnections", 50);          // give up after X number of failed connections.
//    fastfile.set("ipv", 6);                         // explicit ipv4 or ipv6 protocol.
//    fastfile.set("location", "fastfile");           // alternate mod_fastfile location.
//    fastfile.set("chunkSize", 2097152);             // break file into X byte size chunks.
//    fastfile.set("host", "alt.domain.com");         // specify explicit "Host" header. (does not effect what is connected to)
//    fastfile.set("header", "HTTP-Header", "data");  // specify arbitrary HTTP headers. (multiple headers allowed)

    // transfer the file to the server.
    fastfile.send("PATH_TO_LOCAL_FILE_HERE");

    // ...asynchronously monitor fastfile.status (1-200 percentage) and fastfile.complete (enum of error, running, or success) from here...
}
catch (Exception ex) {
    MessageBox.Show("Fastfile fault: " + ex.Message);
}
// -------------------------------------------------------------------------------------------------------------------
*/

using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using System.Text;
using System.Text.RegularExpressions;
using System.IO;

namespace Fastfile
{
    class Fastfile
    {
        private const string ffcpAllowedChars = @"[^a-zA-Z0-9=:. _/\\-]";
        private string ffcpPath = "";
        private int _status = 0; /* contains percentage, 0-200 (0-100 for manifesto building percentage, 100-200 for transfer percentage) */
        public int status { private set { _status = value; } get { return _status; } }
        private string _message = "";
        public string message { private set { _message = value; } get { return _message; } }
        public enum completeStatus { error, running, success }
        private completeStatus _complete = completeStatus.running;
        public completeStatus complete { private set { _complete = value; } get { return _complete; } }
        private BackgroundWorker statusWorker = null;
        private Process proc = null;
        private Dictionary<string, string> client = new Dictionary<string, string>();
        private ArrayList clientHeader = new ArrayList();

        public Fastfile(string path, string hostname, string key) {
            ffcpPath = path;
            client["hostname"] = hostname;
            client["key"] = key;

            /* process the status being streamed to stdout from ffcp.exe (Process.OutputDataReceived proved to be unreliable for this) */
            statusWorker = new BackgroundWorker();
            statusWorker.WorkerReportsProgress = true;
            statusWorker.WorkerSupportsCancellation = true;
            statusWorker.DoWork += new DoWorkEventHandler(statusWorker_DoWork);
            statusWorker.ProgressChanged += new ProgressChangedEventHandler(statusWorker_ProgressChanged);
        }
        ~Fastfile() { stop(); }

        public void stop() {
            try {
                if (proc != null && !proc.HasExited)
                    proc.Kill();
                if (statusWorker != null && statusWorker.IsBusy && !statusWorker.CancellationPending)
                    statusWorker.CancelAsync();
            }
            catch { }
        }

        /* set (optional) ffcp options. (return self to allow chaining of sets) */
        public Fastfile set<T>(string name, T value) {
            switch(name) {
                case "ssl":
                case "location":
                case "bufSize":
                case "timeout":
                case "chunkSize":
                case "connections":
                case "failedConnections":
                case "host":
                case "port":
                case "ipv":
                    client[name] = value.ToString();
                    break;
                case "header":
                    clientHeader.Add(value.ToString());
                    break;
                default:
                    throw new ArgumentException("Invalid set option", name);
            }
            return this;
        }

        /* set (optional) ffcp options. (this is currently only used for "header") (return self to allow chaining of sets) */
        public Fastfile set(string name, string key, string value) {
            set(name, key + "=" + value);
            return this;
        }

        /* send a file to a fastfile server. (using ffcp.exe) */
        public void send(string path) {
            if (proc != null)
                throw new Exception("Process is already active.");

            string arguments = "-q -a"; /* quiet errors and request api-mode output to stdout. */
            Dictionary<string, string> argumentMap = new Dictionary<string, string>();
            argumentMap["ssl"]                  = "-s";
            argumentMap["location"]             = "-l";
            argumentMap["bufSize"]              = "-b";
            argumentMap["timeout"]              = "-t";
            argumentMap["chunkSize"]            = "-c";
            argumentMap["connections"]          = "-n";
            argumentMap["failedConnections"]    = "-f";
            argumentMap["host"]                 = "-h";
            argumentMap["port"]                 = "-p";
            argumentMap["ipv"]                  = "-";
            argumentMap["header"]               = "-H";

            /* don't bother continuing if we can't find the file we're eventually going to send. */
            if (!File.Exists(path))
                throw new FileNotFoundException("Could not find file to send", path);

            /* create command-line argument string. */
            foreach (KeyValuePair<string, string> item in client) {
                if (argumentMap.ContainsKey(item.Key)) {
                    if (item.Key.Equals("ssl")) {
                        if (item.Value.Equals("True"))
                            arguments += " " + argumentMap[item.Key];
                    }
                    else if (item.Key.Equals("ipv")) {
                        if (item.Value.Equals("4") || item.Value.Equals("6"))
                            arguments += " " + argumentMap[item.Key] + item.Value;
                    }
                    else
                        arguments += " " + argumentMap[item.Key] + " \"" + Regex.Replace(item.Value, ffcpAllowedChars, "") + "\"";
                }
            }

            /* add in any specified HTTP header(s). */
            foreach (string header in clientHeader)
                arguments += " " + Regex.Replace(argumentMap["header"], ffcpAllowedChars, "") + " \"" + Regex.Replace(header, ffcpAllowedChars, "") + "\"";

            /* finally, add the hostname, key, and file to the argument list. */
            arguments += " \"" + Regex.Replace(this.client["hostname"], ffcpAllowedChars, "") + "\"";
            arguments += " \"" + Regex.Replace(this.client["key"], ffcpAllowedChars, "") + "\"";
            arguments += " \"" + path.Replace("\"", "") + "\""; /* quotes aren't allowed with windows filenames, just trim it and let it fail. */

            /* setup the process to be in the background and launch. */
            proc = new Process();
            proc.StartInfo.FileName = string.Format(@"{0}\{1}", ffcpPath, "ffcp.exe");
 
            proc.StartInfo.Arguments = arguments;
            proc.StartInfo.WindowStyle = ProcessWindowStyle.Hidden;
            proc.StartInfo.UseShellExecute = false;
            proc.StartInfo.CreateNoWindow = true;
            proc.StartInfo.WorkingDirectory = ffcpPath;
            proc.StartInfo.RedirectStandardOutput = true;
            proc.StartInfo.RedirectStandardError = true;
            proc.StartInfo.StandardOutputEncoding = Encoding.GetEncoding("iso-8859-1"); /* we need to read the 8th bit of a character. */
            proc.Start();
            statusWorker.RunWorkerAsync();
        }

        /* monitor status returns from ffcp.exe. */
        private void statusWorker_DoWork(object sender, DoWorkEventArgs e) {
            int c;
            while (!statusWorker.CancellationPending && !proc.HasExited) {
                try {
                    if ((c = proc.StandardOutput.Read()) >= 0 && status != c)
                        statusWorker.ReportProgress(c);
                }
                catch { }
            }

            /* may have been tripped out of the loop before exiting, kill it if it's still going. */
            if (!proc.HasExited)
                proc.Kill();

            /* really just for the kill above, but it doesn't hurt to have this hit either way. */
            proc.WaitForExit();

            /* record the success/error status. */
            if (statusWorker.CancellationPending)
                message = "User aborted transfer.";
            else if (proc.ExitCode != 0) {
                try { message = proc.StandardError.ReadToEnd(); }
                catch { message = "Unknown error."; }
            }
            else
                message = "Successfully sent file!";

            complete = proc.ExitCode == 0 ? completeStatus.success : completeStatus.error;
        }

        private void statusWorker_ProgressChanged(object sender, ProgressChangedEventArgs e) {
            status = e.ProgressPercentage;
        }
   }
}
